File: encode.py
from dataclasses import dataclass
import string
from typing import IO, Dict

from tqdm import tqdm


rockyou_path = "/home/nabuki/datasets/rockyou.txt"

# Define the character set used in passwords
characters = str(
    string.ascii_lowercase
    + string.ascii_uppercase
    + string.digits
    + string.punctuation,
)
char_to_index = {char: idx for idx, char in enumerate(characters)}


def tokenize(s: str):
    return (char_to_index[ch] for ch in s if ch in characters)


@dataclass
class Node:
    _forward: Dict[int, "Node"]
    _visits: int = 0

    def __init__(self):
        self._forward = {}

    def visit(self, token: int):
        e = self._forward.get(token)
        if e is None:
            e = Node()
            self._forward[token] = e
        e._visits += 1
        return e

    def normalized_transition_weights(self):
        s = sum(node._visits for node in self._forward.values())

        if s == 0:
            return None

        res = [0.0] * len(characters)
        for k, v in self._forward.items():
            res[k] = v._visits / s

        return res

    def write_normalized(self, sink: IO[bytes]):
        transitions = self.normalized_transition_weights()

        # write zeros if endpath
        if transitions is None:
            sink.write(b"\x00" * len(characters))
            return

        # quantize 0..1f to 0..255i
        transitions = [int(v * 255) for v in transitions]

        # write quantized transition probabilities (byte for each token)
        for v in transitions:
            sink.write(v.to_bytes())

        # for each non-zero probability transition, write them recursively
        nonzero_transitions_idexes = [
            index for index, value in enumerate(transitions) if value != 0
        ]
        for i in nonzero_transitions_idexes:
            self._forward[i].write_normalized(sink)


# Collect tree
root = Node()
with open(rockyou_path, "r", encoding="latin1") as file:
    for line in tqdm(file):
        curr = root
        for token in tokenize(line.strip()):
            curr = curr.visit(token)


# Write it as probabilities (normalized)
print("Writing...")
with open("res.pfx", "wb") as f:
    root.write_normalized(f)

File: encode.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <chrono>
#include <thread>

const std::string rockyou_path = "/home/nabuki/datasets/rockyou.txt";

// Define the character set used in passwords
const std::string characters = 
    "abcdefghijklmnopqrstuvwxyz"
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    "0123456789"
    "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";

std::unordered_map<char, int> char_to_index;

std::vector<int> tokenize(const std::string &s) {
    std::vector<int> tokens;
    for (char ch : s) {
        if (char_to_index.find(ch) != char_to_index.end()) {
            tokens.push_back(char_to_index[ch]);
        }
    }
    return tokens;
}

class Node {
public:
    std::unordered_map<int, Node*> _forward;
    int _visits;

    Node() : _visits(0) {}

    Node* visit(int token) {
        if (_forward.find(token) == _forward.end()) {
            _forward[token] = new Node();
        }
        _forward[token]->_visits++;
        return _forward[token];
    }

    std::vector<float> normalized_transition_weights() {
        int s = 0;
        for (const auto& pair : _forward) {
            s += pair.second->_visits;
        }

        if (s == 0) {
            return std::vector<float>();
        }

        std::vector<float> res(characters.size(), 0.0);
        for (const auto& pair : _forward) {
            res[pair.first] = static_cast<float>(pair.second->_visits) / s;
        }

        return res;
    }

    void write_normalized(std::ofstream &sink) {
        auto transitions = normalized_transition_weights();

        if (transitions.empty()) {
            sink.write(std::vector<char>(characters.size(), 0).data(), characters.size());
            return;
        }

        std::vector<char> quantized_transitions(characters.size(), 0);
        std::transform(transitions.begin(), transitions.end(), quantized_transitions.begin(), [](float v) {
            return static_cast<char>(v * 255);
        });

        sink.write(quantized_transitions.data(), quantized_transitions.size());

        for (size_t i = 0; i < quantized_transitions.size(); ++i) {
            if (quantized_transitions[i] != 0) {
                _forward[i]->write_normalized(sink);
            }
        }
    }
};

int main() {
    for (size_t i = 0; i < characters.size(); ++i) {
        char_to_index[characters[i]] = i;
    }

    Node root;
    std::ifstream file(rockyou_path, std::ios::in);
    if (!file.is_open()) {
        std::cerr << "Failed to open file: " << rockyou_path << std::endl;
        return 1;
    }

    std::cout << "Enter the number of lines to process: ";
    size_t max_lines;
    std::cin >> max_lines;

    std::string line;
    size_t line_count = 0;
    auto start_time = std::chrono::steady_clock::now();
    auto last_time = start_time;

    while (std::getline(file, line) && line_count < max_lines) {
        Node* curr = &root;
        auto tokens = tokenize(line);
        for (int token : tokens) {
            curr = curr->visit(token);
        }
        line_count++;

        auto current_time = std::chrono::steady_clock::now();
        auto elapsed_seconds = std::chrono::duration_cast<std::chrono::seconds>(current_time - last_time).count();
        if (elapsed_seconds >= 1) {
            std::cout << "Processed lines: " << line_count << std::endl;
            last_time = current_time;
        }
    }
    file.close();

    std::cout << "Writing..." << std::endl;
    std::ofstream out("res.pfx", std::ios::binary);
    if (!out.is_open()) {
        std::cerr << "Failed to open output file: res.pfx" << std::endl;
        return 1;
    }

    root.write_normalized(out);
    out.close();

    return 0;
}
